### ISSUE TITLE
Finer granularity for the normalization options

### ISSUE THREAD

**Issue by jlstevens** [ioam/holoviews#49]
Right now we have the four normalization options ``+axiswise, -axiswise, +framewise, -framewise``. Although these four modes are already quite flexible as they apply by ``{type}.{group}.{label}`` there are always more possible ways to normalize things!

The most flexible (but verbose!) approach would be to have the four normalization options specified for each dimension by dimension name. Although flexible, this would require a lot more work in the '%%opts' magic parser and would be quite brittle (e.g if you rename your dimensions!).

Having discussed this with Philipp we think there is a useful middle-ground - use the existing split between key dimensions and value dimensions. This would get rid of the ``situated`` parameter for instance, replacing that with the more general notion of normalizing by key dimension.

Here is how it would work in the magic:

``+axiswise, -axiswise, +framewise, -framewise``                   # Current semantics (i.e all dimensions)
``+axiswise[K], -axiswise[K], +framewise[K], -framewise[K]``  # Apply to key dimensions only
``+axiswise[V], -axiswise[V], +framewise[V], -framewise[V]`` # Applyi to value dimensions only

Note that there are equivalences e.g ``-axiswise[K]`` is equivalent to ``+axiswise[V]`` as specifying key 'on' is the same as specifying value 'off'.

This is compatible with the current approach and here is what it would map to as a normalization option as accessed by the plotting code: 

``
axiswise=True|False|'key'|'value'
``

Looking at this, this might be a good opportunity to have the magic syntax match the option more closely. I.e the `+` and `-` prefix are just shortcuts for the boolean options only. I.e the magic would support:

``+axiswise, -axiswise, +framewise, -framewise`` 
``axiswise='key', axiswise='value', framewise='key', framewise='value'`` 

And if you don't want to use the short syntax:

``axiswise=True, axiswise=False, framewise=True, framewise=False`` 

Any thoughts? This shouldn't be **too** much effort to implement, is compatible with what we already have and would certainly be useful in some cases (as well as getting rid of the ``situate`` parameter).

**Comment by jbednar** [comment]
My thoughts:

- The edit addresses my main initial concern, i.e. that there be a backwards-compatible way to specify specific dimensions eventually, even if not used often.  

- I'm not sure how this replaces situate, but if it does, then great.  

- I do support having a closer match between magic and non-magic syntax; I had previously tried axiswise=True and was annoyed that it didn't work.

- I'm a bit unclear about the claimed "equivalences" -- isn't it meaningful to have no normalization at all?  If so, why should turning off one type of normalization turn on another?  This probably just needs some explanation, as I'm finding it difficult to imagine a concrete example that would make it clear.

**Comment by philippjfr** [comment]
The normalization system computes the ranges for both the key (spatial) dimensions, i.e. along the axes, and for value dimensions, e.g. the pixel luminance of an Image. The situate option was introduced for Images to control the relative spatial sizes (just like in the Tk GUI), independently from the value dimension. With this proposal each could be controlled individually, obviating the need for a situate option.

The granularity of this control is what we need to decide now. Having independent control over the key vs. value dimensions already provides a lot more power, however obviously controlling normalization for each dimension either by name or by index would provide the most fine-grained control. We can come up with syntax that is both backward compatible and supports all of these options, however we should decide what level of complexity we want to support, i.e. is the added complexity of providing control over each dimension worth the potential utility or is separate control for key and value dimensions sufficient?

The equivalences arise out of the fact that we have defaults, i.e. setting +axiswise for value dimensions, implicitly says the key_dimensions will be normalized -axiswise, simply because that is the default. Having no normalization at all simply means activating both +axiswise and +framewise, since computing the range per frame and per set of axes is what matplotlib does by default anyway.

**Comment by jbednar** [comment]
Normalization is so confusing!  I would have thought +axiswise and +framewise is *complete* normalization, not no normalization, if it means ranges are computed independently for each plot.  Anyway, it's up to you guys whether to implement the fine-grained control now or later, taking into account (a) you understand it right now, and will gradually forget it if you don't do it now, (b) you will probably need this level of control at some point in your thesis, yet (c) it's distracting to work on now, while you need to get your research work done.  So by default, implement only the bits you need right now, but if you can see how to implement the rest and totally be done with it, and it won't be much distraction, you can do that now.

**Comment by jlstevens** [comment]
I agree that normalization is confusing but for different reasons. If you think of ``+axiswise`` as meaning "normalize per frame" then it (sort of!) make sense right now.

What *is* confusing is that by default HoloViews normalizes *everything* so in order to customize the normalization behavior you have to turn normalization *off* using ``+framewise`` and ``+axiswise`` (i.e instructions to 'only normalize per frame' or per 'only normalize per axis').

I do think that normalization options with `+` prefix with semantics of  *enabling* normalization would be preferable (full normalization still on by default). Currently adding an option with the 'wise' suffix implies the opposite; do *less* normalization instead of 'enable normalization across frames' or 'enable normalization across axes' which is probably a more natural approach.

What might work while also making things *more* succinct is to drop the 'wise' suffix altogether with the convention that you are specifying what to normalize *across*. E.g

```python
%%opts Curve {+frame +axes}
curve
```

This means that normalization is  **enabled** *across* both frames and *across* the axes: the opposite semantics to the current 'wise' suffix equivalents. To reiterate, full normalization will always be   by the default,  so the example above does nothing to change the normal behavior (and hence pointless!).

In practice you **still** would need to switch normalization *off* using the ``-`` prefix  but maybe this approach is (hopefully?) less confusing than using the current ``+wise`` system where the `+` sign is  *disabling* normalization!. For instance:

```python
%%opts Curve {-frame}
curve
```

This is to be read as "switch *off* normalization across frames".  Then "-axes" would mean "switch *off* normalization across axes" and using both would then obviously mean "don't normalize either across axes or across frames". The verbose equivalent is then ``%%opts Curve norm{frame=False, axis=False}``.

If we were to implement the new system proposed above for key/value dimension normalization, I would be happy to implement these new semantics while also supporting the old semantics (backward compatibility). In essence, the presence of the 'wise' suffix just switches the sign around.

Obviously tab-completion would only give suggestions for the new syntax (if you agree it is better!) while allowing the old version to keep working (probably with a deprecation warning though).

Hope that makes more sense now! Unfortunately, normalization is confusing probably because there tons of legitimate and meaningful ways of normalizing data which means the issue will never be trivial to understand...

**Comment by jbednar** [comment]
I think that changing the names at this point would only make it worse!  In any case, I don't see any justification above for treating the - or + options as being *more* or *less* normalization; they all sound like they are always full normalization, just over different quantities.  

Just for comparison, in the Tk GUI, there is a meaningful notion of normalization being turned off.  There, with no normalization, values 0..1.0 in an array are mapped to values 0..255 in an image, with higher values cropped off and the mapping from array to image being entirely independent of the scale of the items in the image.

Here in HoloViews, I don't know of any option like that, where there is an absolute and fixed mapping from array values to image pixel values, not affected by the actual values in the array.  Thus I don't think that any of the options you consider represent an increase or decrease in amount of normalization; they all sound like 100% normalization to me!  The normalization is done with respect to different quantities (hence the -wise suffix), but it's always done, right?  If that's true, I don't think we should change the names, and I don't think any description should talk about normalizing or not normalizing, just explaining what we're normalizing wrt what.  Does that make sense?

**Comment by jlstevens** [comment]
> I don't see any justification above for treating the - or + options as being more or less normalization; they all sound like they are always full normalization, just over different quantities.

You can normalize over more elements or fewer elements. When I say 'more normalization', I mean more things are being normalized together.

> ... values 0..1.0 in an array are mapped to values 0..255 in an image, with higher values cropped off 

That just sounds like a default normalization procedure to me! Mapping 0.0-1.0 to 0..255 is simply a fairly standard convention. To claim there is no normalization, there would be no re-scaling *at all*...

> Here in HoloViews, I don't know of any option like that, where there is an absolute and fixed mapping from array values to image pixel values, not affected by the actual values in the array.

You simply need to define the range on the appropriate value dimension and that will work. As far as the normalization options go, yes...*if* no range is specified on the value dimensions, we scale according to the data.

> I don't think any description should talk about normalizing or not normalizing, just explaining what we're normalizing w.r.t. what. Does that make sense?

I agree but again, you can normalize with respect to more elements or fewer elements (per dimensions) and that is what we are trying to find a way to specify nicely.

**Comment by jbednar** [comment]
Good point; I'm distinguishing between rescaling (any operation that maps from one value range to another) and normalization (a rescaling operation that takes the actual data values into account, not just a fixed range).  I tried looking at various definitions of normalization online just now, and didn't find evidence for or against such a distinction.  

I checked what matlab calls it, in case it clarified it, but as you might expect that just makes it worse -- matlab's image() function does what I call a non-normalized mapping (fixed scaling independent of data values), while imagesc() (for "scale") does what I call a normalized mapping (based on the data values).

Not sure where else to check for possible terms to use, so I've no idea what to call anything.

**Comment by jbednar** [comment]
Ok, I did find an example of the sense in which I was using normalization:

  http://en.wikipedia.org/wiki/Normalization_%28image_processing%29

That's clearly what I've been meaning for topo.plotting and the Tk GUI, which makes sense because those are both based explicitly on visualizing images.  But HoloViews shows things as plots, not images per se, so it's not precisely the same case and thus it's not as clear that this image-processing definition of normalization applies.

**Comment by philippjfr** [comment]
I've just been thinking about this a bit more and I think I know how to replace the unclear framewise/axiswise options, really you want to control across what kind of containers something is normalized and what dimensions are actually normalized. I'd propose that you'd declare something like:

```python
%%opts Image {holomap=False layout=['z']}
```

In this scheme ``axiswise`` is equivalent to ``layout=False`` and ``framewise`` equivalent to ``holomap=False``. This would additionally allow specifying whether things are normalized across a grid or not. Secondly instead of just accepting boolean values you can specify which dimensions to normalize over either by referring to the dimension by name or by an integer corresponding to their position in the Element kdims and vdims. This achieves everything we've discussed with what I think is a very clear syntax.

**Comment by jbednar** [comment]
Sounds good to me. I've always thought that we needed to be able to say explicitly what should get normalized together, both for power and for ease of understanding.  Here you've used a list for 'z' because there could be more than one dimension normalized, e.g. `layout=['x','z']`, to normalize the x and z axes together (but not the y axis)?  

And in that example presumably it's clear that x will be normalized with x in other plots, not together with z somehow?  For x and z combining them seems nonsensical, but e.g. for `layout=['x','y']` maybe it isn't -- combining them would presumably find the smallest square area enclosing the range of x and y values?  So are you meaning to include or exclude this possibility?

BTW, Jean-Luc mentions above that referring to a dimension by name was more brittle than by index, but it seems *more* robust to me, i.e. less likely to cause undetected errors, since the name is meaningful to a human.

**Comment by philippjfr** [comment]
>And in that example presumably it's clear that x will be normalized with x in other plots, not together with z somehow?

Dimensions with different names will never be normalized together unless they occupy the same axis in an overlay plot. By including/omitting particular dimensions you're basically saying normalize or don't normalize this dimension across this container, not whether to normalize across them. By default all the dimensions of an Element are normalized by declaring ``layout=['z']`` you are implicitly saying don't normalize x and y. Perhaps it would be worthwhile to add a keyword to explicitly denote to normalize all dimensions, maybe ``layout='all'``.

>but e.g. for layout=['x','y'] maybe it isn't -- combining them would presumably find the smallest square area enclosing the range of x and y values? So are you meaning to include or exclude this possibility?

This is an interesting idea but I'm not sure how you'd express it in this scheme.

>BTW, Jean-Luc mentions above that referring to a dimension by name was more brittle than by index, but it seems more robust to me, i.e. less likely to cause undetected errors, since the name is meaningful to a human.

It depends on what you are wanting to do, if you want to normalize the x/y-axis of a plot then you can just set ``layout=[0,1]`` without thinking in detail about the dimensions of the plot, while referring to a dimension by name can be useful if you've got a lot of plots with different values and you want to change the normalization for a specific one.

**Comment by jbednar** [comment]
> >but e.g. for layout=['x','y'] maybe it isn't -- combining them would presumably find the smallest square area enclosing the range of x and y values? So are you meaning to include or exclude this possibility?

>This is an interesting idea but I'm not sure how you'd express it in this scheme.

I was mainly just making sure I understood the semantics you were intending.  I guess if we did want to support pooling axes together for normalization, it could be expressed similarly: `layout=[['x','y'],'z']`.

**Comment by jbednar** [comment]
> Edit: We could also allow specific keywords like 'xaxis', 'yaxis', 'color', 'size', 'angle', 'magnitude' and 'alpha' which might be a little clearer than specifying them by their integer position or even by name.

If you supported that, seems like one would want to be able to specify both True and False for each of the independently, not to have ones default to False if not listed as True, or vice versa.  That would make the syntax longer: `layout={xaxis:True,yaxis:False}` (where z defaults to whatever it defaults to), but seems clearer to me.

**Comment by philippjfr** [comment]
Simple pooling, i.e. just finding the min/max across them wouldn't be very useful though, might make more sense in combination with an option to normalize the range not the overall values.

**Comment by philippjfr** [comment]
>If you supported that, seems like one would want to be able to specify both True and False for each of the independently, not to have ones default to False if not listed as True, or vice versa. That would make the syntax longer: layout={xaxis:True,yaxis:False} (where z defaults to whatever it defaults to), but seems clearer to me.

I think this scheme might actually be clearer in general since it's more explicit, since it defaults to True you'd usually only have to set the False values.

**Comment by jbednar** [comment]
> Simple pooling, i.e. just finding the min/max across them wouldn't be very useful though, might make more sense in combination with an option to normalize the range not the overall values.

Right.  I very often do want to make the ranges (i.e. areas) covered by the plots to be similar, even if the numeric values differ in each plot.  Not sure if I want to do that across *both* axes together, but maybe...

**Comment by philippjfr** [comment]
Really if we also want to provide control over that then we'd need a scheme like ``layout={xaxis: area, yaxis: area, color: range}``.

**Comment by jbednar** [comment]
That seems even more explicit and more powerful.  Is there any other meaningful option besides area and range?  

The word 'area' is misleading here, in any case; when applied to a single axis it's not an area (a 2D-specific concept) but a linear distance or length.  'extent' probably would have worked, but you're already using that to mean an absolute region of the axis, not the length of that region, right?  Is 'range' consistently used to mean an absolute region of an axis, in hv code?  In the real world, people use it to mean either one, i.e. the specific start and end values and the difference between the end and the start.

**Comment by philippjfr** [comment]
Yea, I had trouble coming up with a name for it. In HoloViews the range is the min and max value, and we already use extents to specify overrides for the axis ranges on an Element, so it's a bit difficult.

**Comment by jlstevens** [comment]
Ok, I'm catching up here. I'll note that this issue is nearly a year old. :-)

From the sound of it, the new suggestion also retains the overall concepts corresponding to 'axiswise' and 'framewise' but with a more detailed specification of which dimensions are normalized and how.

I'm not *entirely* sold on the idea of enumerating keywords although I do think it is nicer than integer indices. If I understand it right, these keywords would also eliminate the need to use dimension names which is another bonus. The settings would depend on the plot type and nothing else which sound great if the following two conditions hold:

1. Is 'xaxis', 'yaxis', 'color', 'size', 'angle', 'magnitude' and 'alpha' a complete list? Or a nearly complete list? How long would a complete list be? If a short, static list is sufficient then it could be tab-completable which would be great!

2. Is there a consistent mapping from these keywords to dimension index? If not, how much code is there to map these keywords to dimension names? I imagine the mapping could be declared as a class attribute on the plotting classes but I believe there will still be more complicated cases. For instance, with ``Points`` we can map the value dimension to different visual attributes. I'm just wondering how much extra code would be needed to support these keywords in the plotting code...

As for the types of normalization, I think the 'range' suggestion makes sense (as the range method is actually used) and I agree that ``extents`` would be a confusing name. So how about ``region``? I don't think we've used that word yet and the concept generalizes to any number of dimensions (except 0D I suppose).

Lastly, I would keep the names ``axiswise`` and ``framewise`` to be consistent. I'm not sure ``layout`` and ``holomap`` are much clearer than what we already have. Combining these new suggestions, would the new syntax be something like this?:

```
%%opts Points { axiswise={xaxis: region, yaxis: area, color: range} }
```
 
There is still a few of things I need a bit longer to think about. For instance, are there contradictory settings? I assume these keywords can't be specified to be axiswise *and* mapwise at the same time. I would also want to think about the magic syntax: the above suggestion uses something that isn't quite a valid Python dictionary.

**Comment by philippjfr** [comment]
>Is 'xaxis', 'yaxis', 'color', 'size', 'angle', 'magnitude' and 'alpha' a complete list? Or a nearly complete list? How long would a complete list be? If a short, static list is sufficient then it could be tab-completable which would be great!

I think that's pretty much complete the only thing I can think of that I missed was zaxis.

>Is there a consistent mapping from these keywords to dimension index? If not, how much code is there to map these keywords to dimension names? I imagine the mapping could be declared as a class attribute on the plotting classes but I believe there will still be more complicated cases. For instance, with Points we can map the value dimension to different visual attributes. I'm just wondering how much extra code would be needed to support these keywords in the plotting code...

No as you point out the mapping is not consistent for certain types so the normalization code would have to look up the plot options as well. I'd likely couple this with my proposal in #581 and register functions or classmethods which know how to compute these attributes.

>As for the types of normalization, I think the 'range' suggestion makes sense (as the range method is actually used) and I agree that extents would be a confusing name. So how about region? I don't think we've used that word yet and the concept generalizes to any number of dimensions (except 0D I suppose).

Region seems like a good suggestion.

>Lastly, I would keep the names axiswise and framewise to be consistent. I'm not sure layout and holomap are much clearer than what we already have. Combining these new suggestions, would the syntax be something like this?:

I would not, if you read some discussions even we frequently confuse these concepts and I'm quite sure that Jim still thinks they mean the opposite to what they actually do. That's simply to say that they are incredibly confusing. It seems much more appropriate for us to declare which containers to normalize over. For backward compatibility we would keep them but I would like to ditch them asap.

>There is still a few of things I need a bit longer to think about. For instance, are there contradictory settings?

No, there's potentially weird settings but every combination of these options can be computed.

**Comment by jlstevens** [comment]
> I would not, if you read some discussions even we frequently confuse these concepts and I'm quite sure that Jim still thinks they mean the opposite to what they actually do.

I agree it is confusing but it is what we've been using so far. Although we spent a *lot* of time trying to come up with those names, I do want something clearer!

Using ``layout`` and ``holomap`` as fixed keywords in this particular way strikes me as rigid as it uses class names in a fixed way, The ``%%opts`` magic is already designed to assign any group of settings to any collection of dynamically registered classes flexibly and dynamically. Therefore, to be consistent with how the magic works in general (i.e for styles and plot options on elements),  something like this would feel more in keeping with the general philosophy:

```
%%opts HoloMap { Points : [xaxis:region, color:range] }
```

* The normalization options (i.e syntax in braces) would be supported by container types.
* Unlike the fixed ``layout`` and ``holomap`` keyword suggestion, the style system would remain completely separate from any details about elements, including their names. The style system only relates to elements once they are registered.
* The options such as ``xaxis:region`` would tab-complete from a fixed list of 14 elements (7 plotting properties x 2 normalization types).

This particular suggestion would group all normalization options for a particular container class in one place instead of  distributing them over the element settings. As this has advantages and disadvantages and my real objection is to any fixed, lowercase keywords that mirror 'special' class names, I would also be happy with the distributed organization:

```
%%opts Points { HoloMap  [xaxis:region, color: range] } 
```

Here ``HoloMap`` would be one tab-completable option from all available container types (which is also a far shorter list than the list of all available elements).

This is then very similar to your original suggestion but retains extensibility by keeping the actual option settings separate from a fixed list of 'special' class names. From these examples, it is clear that normalization is defined by a joint relationship between element types, container types and plotting dimensions for the corresponding plot type. In other words, you either need to select from a dynamic list of container types and then element types (i.e the first suggestion) or the other way round.

On balance, I prefer the second suggestion. It is closer to what we already use and looks very similar to your suggestion above. What do you think?

**Comment by philippjfr** [comment]
I like the second suggestion, but there is a reason why I suggested the lower case version before. Basically we have equivalent types like NdLayout/Layout, GridSpace/GridMatrix, which aren't nestable and therefore do not require their own keywords, I don't particularly object to controlling them separately but my original thought was that ``layout`` would refer to both types of ``Layout`` and ``grid`` would refer to both types of ``Grid``.

**Comment by jlstevens** [comment]
That seems to be the issue - we want a concept that means something across classes but can't find a name for what we mean!

Either we come up with great meaningful names that are clearer than axiswise/framewise or we use actual class names to refer to settings of those actual classes. Using class names to try to point to a nebulous concept of 'layout-like' and 'frame-like' doesn't feel right to me.

As for avoiding entries for things like ``GridSpace``, I think we could suppress containers that *only* have normalization settings from the rest of the magic and tab-complete from the list of containers with normalization options inside the normalization options block (i.e inside the braces).

**Comment by philippjfr** [comment]
>As for avoiding entries for things like GridSpace, I think we could suppress containers that only have normalization settings from the rest of the magic and tab-complete from the list of containers with normalization options inside the normalization options block (i.e inside the braces).

I can't parse that. We do want to be able to control normalization across a GridSpace. I've come around to your suggestion now and do think we should refer to the container classes by name, which would include HoloMap, Overlay, NdOverlay, NdLayout, Layout, GridSpace and GridMatrix.

**Comment by jlstevens** [comment]
I mean that when you press tab here:

```
%%opts [TAB]
```

Then ``GridSpace`` doesn't come up if there are no applicable style/plot options for it. But if you press tab here:

```
%%opts Points { [TAB]
```

Then ``GridSpace`` is listed to specify normalization options on GridSpace.

Hope that is clearer!

**Comment by philippjfr** [comment]
But GridSpace does have plot options that can be customized so it should tab-complete in both cases.

**Comment by jlstevens** [comment]
GridSpace was a bad example then. I'm just thinking about how to handle potential containers that only have normalization options as you wouldn't want them listed after ``%%opts [TAB]``.

This is just a minor point anyway and I think I'm finally happy with the overall suggestion!

**Comment by philippjfr** [comment]
>I'm just thinking about how to handle potential containers that only have normalization options as you wouldn't want them listed after %%opts [TAB]

All containers already tab-complete and all containers have at least some plot options that can be applied.

**Comment by jbednar** [comment]
> Region seems like a good suggestion.

To me, "region" is a concept that applies by default to 2D and higher dimensions; for a line it seems unnatural (not nonsensical, but unnatural), to say "region".   And even if I did, I don't think it conveys "length" -- if I wanted to talk about a region of a number line, I'd expect to mean e.g. the part between 7.2 and 9.8; whereas the difference (2.6) is not a region but a distance.  The original suggestion was 'area', which is very clear in 2D yet is actually being applied one dimension at a time, and the precise correlate of area for 1D is length, right?  So I think normalizing the x axis length across plots is clear -- not the specific range, but just the length of it.

> if you read some discussions even we frequently confuse these concepts and I'm quite sure that Jim still thinks they mean the opposite to what they actually do. 

Actually, I've simply resigned myself to just trying all four combinations whenever it's not doing what I want.  That's probably just because of history, though -- I was permanently scarred by the process of coming up with those names, getting confused by the previous suggestions, and was never able to embrace the final syntax fully.  Obviously,  with a fully powerful system there will be more than four combinations and I'll need to figure it out, so I want to make sure we have a good, clear system this time!

> For backward compatibility we would keep them but I would like to ditch them asap.

Yes, please.  Keep for backwards compatibility, but purge all uses of them from our materials and eventually delete them.

> All containers already tab-complete and all containers have at least some plot options that can be applied.

I would think that even if normalization options are the only options that could be set on some container type, we'd still want it to tab complete; normalization options are options, just like any other.

> ```python
%%opts HoloMap { Points : [xaxis:region, color:range] }
%%opts Points { HoloMap  [xaxis:region, color: range] } 
```

I'm not sure which of these I prefer, but first, would the .group.label syntax work for both of them?  I.e., would we be able to specify a normalization pool specifically for only things matching Points.Group1?  If we can achieve that then I think we'll have full power over all meaningful normalizations, but I'm not sure what would happen to the Points that don't match Points.Group1.  Do we need an explicit concept of a normalization pool, saying that by default all objects of the same type in a given container form a single normalization pool, but by specifying a normalization option per group or per group.label then you've created a new normalization pool, which is then normalized together (if the options chosen enable that) while the remaining Points objects continue to be normalized with each other as before?

> It probably won't be implemented for another few months but having a plan was always the biggest obstacle.

I'd hope that the project where normalizing by area came up will be starting up again in a couple of weeks, but it's not necessarily the first thing we have to address...
